cloudwatch_logs_common/src/log_publisher.cpp

```diff
@@ -108,102 +107,153 @@ Aws::CloudWatchLogs::ROSCloudWatchLogsErrors LogPublisher::StopPublisherThread()
+  auto shared_logs_obj = this->shared_logs_.load(std::memory_order_acquire);
+  if (nullptr != shared_logs_obj) {
+    std::list<Aws::CloudWatchLogs::Model::InputLogEvent> * logs;
+    publisher_status = shared_logs_obj->getDataAndLock(logs);
+    if (CW_LOGS_SUCCEEDED == publisher_status) {
+      if (!logs->empty()) {
+        Aws::CloudWatchLogs::ROSCloudWatchLogsErrors send_logs_status = CW_LOGS_FAILED;
+        int tries = kMaxRetries;
+
+        while (CW_LOGS_SUCCEEDED != send_logs_status && tries > 0) {
+          send_logs_status = this->cloudwatch_facade_->SendLogsToCloudWatch(
+            next_token, this->log_group_, this->log_stream_, logs);
           if (CW_LOGS_SUCCEEDED != send_logs_status) {
-            AWS_LOG_WARN(
-              __func__,
-              "Unable to send logs to CloudWatch and retried, dropping this batch of logs ...");
+            AWS_LOG_WARN(__func__, "Unable to send logs to CloudWatch, retrying ...");
+            Aws::CloudWatchLogs::ROSCloudWatchLogsErrors get_token_status =
+              this->cloudwatch_facade_->GetLogStreamToken(this->log_group_, this->log_stream_,
+                                                          next_token);
+            if (CW_LOGS_SUCCEEDED != get_token_status) {
+              AWS_LOG_WARN(__func__,
+                           "Unable to obtain the sequence token to use, quit attempting to send "
+                           "logs to CloudWatch");
+              break;
+            }
           }
+          tries--;
+        }
+        if (CW_LOGS_SUCCEEDED != send_logs_status) {
+          AWS_LOG_WARN(
+            __func__,
+            "Unable to send logs to CloudWatch and retried, dropping this batch of logs ...");
         }
       }
-      // TODO: For now we're just going to discard logs that fail to get sent. Later we may add some
-      // retry strategy
-
-      /* Null out the class reference before unlocking the object. The external thread will be
-         looking to key off if the shared object is locked or not to know it can start again, but
-         the PublishLogs function checks if the the pointer is null or not. */
-      this->shared_logs_.store(nullptr, std::memory_order_release);
-      shared_logs_obj->freeDataAndUnlock();
     }
-    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+    // TODO: For now we're just going to discard logs that fail to get sent. Later we may add some
+    // retry strategy
+
+    /* Null out the class reference before unlocking the object. The external thread will be
+       looking to key off if the shared object is locked or not to know it can start again, but
+       the PublishLogs function checks if the the pointer is null or not. */
+    this->shared_logs_.store(nullptr, std::memory_order_release);
+    shared_logs_obj->freeDataAndUnlock();
+  }
+}
+
+void LogPublisher::Run()
+{
+  Aws::String next_token;
+  this->run_state_ = LOG_PUBLISHER_RUN_CREATE_GROUP;
```

Comment:
You could run them sequentially, but each one would then need its own loop to continue until it was successful. This pattern also sets us up for handling more complex situations in the future like if you were trying to send logs and someone deleted your stream or group. In that case you would get an exception back from the service telling you that the group or stream doesn't exist and you can easily kick back into the state for recreating those and continue on your way instead of crashing or failing forever.